File: ios/StrainWise/Doctors/DoctorsModel.swift
import CoreLocation
import Foundation

@Observable
@MainActor
final class DoctorsModel {
    enum Status: Equatable {
        case idle
        case locating
        case searching
        case ready
        case error
    }

    var status: Status = .idle
    var errorMessage: String?
    var result: DoctorResult?
    var radiusMiles: Double = 25
    var city: String = ""
    var state: String = ""
    var lastResolvedLocation: DoctorResolvedLocation?

    @ObservationIgnored private let api: any StrainServicing
    @ObservationIgnored private let location: LocationProvider

    init(
        api: any StrainServicing = LiveStrainAPI(),
        location: LocationProvider? = nil,
    ) {
        self.api = api
        self.location = location ?? LocationProvider()
    }

    func useMyLocation() async {
        status = .locating
        errorMessage = nil
        do {
            let coords = try await location.requestLocation()
            await runSearch(query: DoctorQuery(
                lat: coords.coordinate.latitude,
                lon: coords.coordinate.longitude,
                radiusMiles: radiusMiles,
            ))
        } catch let error as LocationProvider.LocationError {
            status = .idle
            errorMessage = error.errorDescription
        } catch {
            status = .idle
            errorMessage = error.localizedDescription
        }
    }

    func searchByCity() async {
        let city = city.trimmingCharacters(in: .whitespacesAndNewlines)
        let state = state.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !city.isEmpty, !state.isEmpty else {
            errorMessage = "Enter a city and state."
            return
        }
        await runSearch(query: DoctorQuery(
            city: city,
            state: state,
            radiusMiles: radiusMiles,
        ))
    }

    func runSearch(query: DoctorQuery) async {
        status = .searching
        errorMessage = nil
        do {
            let found = try await api.findDoctors(query: query)
            result = found
            if let resolved = found.resolvedLocation {
                lastResolvedLocation = resolved
                city = resolved.city
                state = resolved.state
            }
            status = .ready
        } catch {
            errorMessage = error.localizedDescription
            status = .error
        }
    }

    static var previewEmpty: DoctorsModel {
        let model = DoctorsModel(api: PreviewStrainAPI())
        return model
    }

    static var previewLoaded: DoctorsModel {
        let model = DoctorsModel(api: PreviewStrainAPI())
        model.result = DoctorResult(
            doctors: [.sample],
            resolvedLocation: DoctorResolvedLocation(city: "Denver", state: "CO", lat: 39.74, lon: -104.99),
            source: "leafly.com/medical-marijuana-doctors/colorado/denver",
        )
        model.city = "Denver"
        model.state = "CO"
        model.status = .ready
        return model
    }
}

File: android/app/src/main/java/com/strainwise/app/ui/doctors/DoctorsModel.kt
package com.strainwise.app.ui.doctors

import com.strainwise.app.StrainWiseApplication
import com.strainwise.app.data.StrainAPI
import com.strainwise.app.models.DoctorQuery
import com.strainwise.app.models.DoctorResolvedLocation
import com.strainwise.app.models.DoctorResult
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * Doctors view-model. Direct port of the iOS `DoctorsModel`.
 *
 *  - [status] is one of idle / locating / searching / ready / error.
 *  - [city] / [state] / [radiusMiles] feed the city-search path.
 *  - [result] / [lastResolvedLocation] are the most recent
 *    successful search response.
 */
class DoctorsModel(
    private val api: StrainAPI = StrainWiseApplication.strainAPI,
    private val locationProvider: LocationProvider? = null,
) {
    enum class Status { Idle, Locating, Searching, Ready, Error }

    private val _status = MutableStateFlow(Status.Idle)
    val status: StateFlow<Status> = _status.asStateFlow()

    private val _errorMessage = MutableStateFlow<String?>(null)
    val errorMessage: StateFlow<String?> = _errorMessage.asStateFlow()

    private val _result = MutableStateFlow<DoctorResult?>(null)
    val result: StateFlow<DoctorResult?> = _result.asStateFlow()

    private val _radiusMiles = MutableStateFlow(25.0)
    val radiusMiles: StateFlow<Double> = _radiusMiles.asStateFlow()

    private val _city = MutableStateFlow("")
    val city: StateFlow<String> = _city.asStateFlow()

    private val _state = MutableStateFlow("")
    val state: StateFlow<String> = _state.asStateFlow()

    private val _lastResolvedLocation = MutableStateFlow<DoctorResolvedLocation?>(null)
    val lastResolvedLocation: StateFlow<DoctorResolvedLocation?> = _lastResolvedLocation.asStateFlow()

    fun setCity(value: String) { _city.value = value }
    fun setState(value: String) { _state.value = value }
    fun setRadius(value: Double) { _radiusMiles.value = value }

    private val location: LocationProvider by lazy {
        locationProvider ?: LocationProvider(StrainWiseApplication.let { it as android.content.Context })
    }

    suspend fun useMyLocation() {
        _status.value = Status.Locating
        _errorMessage.value = null
        try {
            val coords = location.requestLocation()
            runSearch(
                DoctorQuery(
                    lat = coords.latitude,
                    lon = coords.longitude,
                    radiusMiles = _radiusMiles.value,
                ),
            )
        } catch (e: LocationProvider.LocationError) {
            _status.value = Status.Idle
            _errorMessage.value = e.message
        } catch (t: Throwable) {
            _status.value = Status.Idle
            _errorMessage.value = t.localizedMessage
        }
    }

    suspend fun searchByCity() {
        val city = _city.value.trim()
        val state = _state.value.trim()
        if (city.isEmpty() || state.isEmpty()) {
            _errorMessage.value = "Enter a city and state."
            return
        }
        runSearch(DoctorQuery(city = city, state = state, radiusMiles = _radiusMiles.value))
    }

    private suspend fun runSearch(query: DoctorQuery) {
        _status.value = Status.Searching
        _errorMessage.value = null
        try {
            val found = api.findDoctors(query = query)
            _result.value = found
            found.resolvedLocation?.let { resolved ->
                _lastResolvedLocation.value = resolved
                _city.value = resolved.city
                _state.value = resolved.state
            }
            _status.value = Status.Ready
        } catch (t: Throwable) {
            _errorMessage.value = t.localizedMessage ?: "Couldn't search."
            _status.value = Status.Error
        }
    }
}
